Mathlib/Control/Monad/Free.lean

```diff
@@ -0,0 +1,260 @@
+  id_map := id_map
+  comp_map _ _ := comp_map _ _
+
+instance : Monad (FreeM F) where
+
+instance : LawfulMonad (FreeM F) := LawfulMonad.mk'
+  (bind_pure_comp := bind_pure_comp)
+  (id_map := id_map)
+  (pure_bind := pure_bind)
+  (bind_assoc := FreeM.bind_assoc)
+
+section liftM
+variable {m : Type u → Type w} [Monad m] {α β : Type u}
+
+/--
+Interpret a `FreeM f` computation into any monad `m` by providing an interpretation
+function for the effect signature `f`.
+
+This function defines the *canonical interpreter* from the free monad `FreeM f` into the target
+monad `m`. It is the unique monad morphism that extends the effect handler
+`interp : ∀ {β}, F β → m β` via the universal property of `FreeM`.
+-/
+protected def liftM (interp : {ι : Type u} → F ι → m ι) : FreeM F α → m α
+  | .pure a => pure a
+  | .liftBind op cont => interp op >>= fun result => (cont result).liftM interp
+
+@[simp]
+lemma liftM_pure (interp : {ι : Type u} → F ι → m ι) (a : α) :
+    (.pure a : FreeM F α).liftM interp = pure a := rfl
+
+@[simp]
+lemma liftM_liftBind (interp : {ι : Type u} → F ι → m ι) (op : F β) (cont : β → FreeM F α) :
+    (liftBind op cont).liftM interp = (do let b ← interp op; (cont b).liftM interp) := by
+  rfl
+
+@[simp]
+lemma liftM_bind [LawfulMonad m]
+    (interp : {ι : Type u} → F ι → m ι) (x : FreeM F α) (f : α → FreeM F β) :
+    (x.bind f : FreeM F β).liftM interp = (do let a ← x.liftM interp; (f a).liftM interp) := by
+  induction x generalizing f with
+  | pure a => simp only [pure_bind, liftM_pure, LawfulMonad.pure_bind]
+  | liftBind op cont ih =>
+    rw [FreeM.bind, liftM_liftBind, liftM_liftBind, bind_assoc]
+    simp_rw [ih]
+
+/--
+A predicate stating that `g : FreeM F α → m α` is an interpreter for the effect
+handler `f : ∀ {α}, F α → m α`.
+
+This means that `g` is a monad morphism from the free monad `FreeM F` to the
+monad `m`, and that it extends the interpretation of individual operations
+given by `f`.
+
+Formally, `g` satisfies the two equations:
+- `g (pure a) = pure a`
+- `g (liftBind op k) = f op >>= fun x => g (k x)`
+-/
+structure Interprets (f : {ι : Type u} → F ι → m ι) (g : FreeM F α → m α) : Prop where
+  apply_pure (a : α) : g (pure a) = pure a
+  apply_liftBind {ι : Type u} (op : F ι) (cont : ι → FreeM F α) :
+    g (liftBind op cont) = f op >>= fun x => g (cont x)
+
+theorem Interprets.eq
+    {f : {ι : Type u} → F ι → m ι} {g : FreeM F α → m α} (h : Interprets f g) :
+    g = (·.liftM @f) := by
+  ext x
+  induction x with
+  | pure a => exact h.apply_pure a
+  | liftBind op cont ih =>
+    rw [liftM_liftBind, h.apply_liftBind]
+    simp [ih]
+
+theorem interprets_liftM (f : {ι : Type u} → F ι → m ι) :
+    Interprets f (·.liftM f : FreeM F α → _) where
+  apply_pure _ := rfl
+  apply_liftBind _ _ := rfl
+
+/--
+The universal property of the free monad `FreeM`. That is, `liftM f` is the unique interpreter that
+extends the effect handler `f` to interpret `FreeM F` computations in monad `m`.
+-/
```

Comment:
I had a go at writing this the "normal" way for universal properties, but it ended up fairly annoying:
```

def liftMEquiv [LawfulMonad m]:
    ({ι : Type u} → F ι → m ι) ≃
      {g : {α : Type u} → FreeM F α → m α // LawfulMonadLift (inst := ⟨@g⟩)} where
  toFun f := .mk (·.liftM f) <| letI : MonadLift (FreeM F) m := ⟨(·.liftM f)⟩; ⟨
    fun _ => by simp,
    fun x f => by cases x <;> simp [FreeM.bind]⟩
  invFun g _ op := g.1 (.lift op)
  left_inv _ := by simp
  right_inv g := by
    letI : MonadLift (FreeM F) m := ⟨g.1⟩
    obtain ⟨g, hg⟩ := g
    have {α} : @g α = monadLift := rfl
    ext _ x
    cases x
    · simp [this]
      simp [← pure_eq_pure]
    · simp [this]
      sorry
```
something to revisit in a later PR I think.